Collection of small fixes in Actor files

sukibaby opened the pull request:
A collection of smaller commits not worth individual PR's mainly concerning improvement of switches (for readability, or for performance) in Actor* code, or stuff heavily used by it..

The commits in this PR are:

1. **Change `ARROW_SIZE` (defined in ScreenDimensions.h) from a macro to a constexpr**
    - Used repeatedly in various switch cases targeted by this PR.
  
2. **ActorFrame: replace a switch with an if/else**
    - More of a stylistic change than an optimization, but makes it easier to understand what's happening.
    
3. **Rename the `StretchType` enum in Actor**
    - The lowercase enum is confusing because it looks like a variable. This prefixes it with a class name so it's easier to understand its purpose.
    
4. **Indicate fallthrough to prevent a GCC warning in ActorMultiVertex**
    - Without this here, GCC 11 and 12 will print a warning during compilation about a possible fallthrough, but we're doing that on purpose, so this simply adds an indicator to the compiler that we know what we're doing. The ifdef's are necessary because, annoyingly, MSVC / XCode may complain if we put the [[fallthrough]] here.

sukibaby committed:
Change ARROW_SIZE to a constexpr int

jsirex commented on src/Actor.cpp:
```diff
@@ -498,6 +498,17 @@ void Actor::PostDraw() // reset internal diffuse and glow
+	constexpr float ONE_HALF = 0.5f;
+	constexpr float ONE = 1.0f;
+	constexpr float TWO = 2.0f;
+	constexpr float TWO_PI_OVER_THREE = 3.1415926536f / 3.0f;
```
ONE_PI_OVER_THREE?

sukibaby replied:
good catch, appreciate it. Though I realized this should probably be in its own PR, even though it's related to improving the switch function.

aeubanks commented:
The compiler is typically smart enough to optimize switch statements better than a human, converting switches to ifs is unlikely to be beneficial (unless you can show an assembly change and demonstrate it's better)

sukibaby commented:
> The compiler is typically smart enough to optimize switch statements better than a human, converting switches to ifs is unlikely to be beneficial (unless you can show an assembly change and demonstrate it's better)

I agree that changing the small switch cases to if-else's likely doesn't provide any performance gain, but in the past when encountering these, @teejusb mentioned a preference to not use switch cases where they're not needed, so I've already replaced a few simple switch cases with if-else.

I changed the small switch cases in `Actor` and `ActorFrame` in anticipation of it being preferred stylistically for this reason.

aeubanks commented:
if changing switches to ifs is done for stylistic reasons that's fine, but the commit description seems to imply that this is done for optimization reasons:

`Replace switch cases with more efficient methods`

sukibaby commented:
> if changing switches to ifs is done for stylistic reasons that's fine, but the commit description seems to imply that this is done for optimization reasons:
> 
> `Replace switch cases with more efficient methods`

I will be happy to amend this to be clear that those two if-else loops were changed for stylistic reasons. (I will wait to push this description change into a commit until this has been reviewed though).

teejusb commented:
> @teejusb mentioned a preference to not use switch cases where they're not needed, so I've already replaced a few simple switch cases with if-else.

The preference just depends on what it's doing. I'm not against switches by any means but in some cases if statements are just more readable.

teejusb commented on src/Actor.cpp:
```diff
@@ -1078,14 +1078,13 @@ void Actor::ScaleTo( const RectF &rect, StretchType st )
 
 	float fNewZoom = 0.f;
-	switch( st )
+	if (st == cover)
```
For enums I actually do prefer switches. I think it would be worth to:
1) Prefix the cases with the enum type e.g. `StretchType::Cover`
2) Rename the enum with better casing. Style guide says to name it similar to constants (i.e UpperCamelCase with a `k` prefix, but I think for now we don't need the `k` prefix. So just `Cover` and `FitInside`. All lowercase is a bit confusing because I thought it was a variable name originally.

teejusb commented on src/ActorFrame.cpp:
```diff
@@ -413,17 +413,17 @@ void ActorFrame::PushChildTable(lua_State* L, const RString &sName)
 		if(a->GetName() == sName)
 		{
-			switch(found)
+			if (found == 0)
```
This is good! IMO this makes way more sense to be an if statement like you have now.

Having it be a switch seems weird when we're doing arithmetic/logic with the variable.

teejusb commented on src/ActorMultiVertex.cpp:
```diff
@@ -289,9 +291,34 @@ void ActorMultiVertex::DrawPrimitives()
 
-void ActorMultiVertex::DrawInternal( const AMV_TweenState *TS )
-{
+using DrawFunction = std::function<void(const RageSpriteVertex*, int, float)>;
```
I do think this is less legible and that the switch is actually preferable. For relatively small and dense cases (like in the case of enums), jump tables are often used which are pretty efficient.

These global maps also aren't trivially destructible so this breaks the style guide.

Here and everywhere below.

teejusb commented on src/ActorMultiVertex.cpp:
```diff
@@ -448,138 +449,154 @@ void ActorMultiVertex::SetSecondsIntoAnimation(float seconds)
-				const auto& rect = _states[state_id].rect;
-
-				switch ((i - first) % 4)
+		using SetVertexTexCoordsFunction = void(*)(std::vector<RageSpriteVertex>&, std::size_t, std::size_t, const std::vector<std::size_t>&, const std::vector<State>&, int);
```
Switch is fine here as well. This is very hard to read.

sukibaby committed:
ActorFrame: replace a switch with an if/else

More of a stylistic change than an optimization, but makes it easier to understand what's happening.

sukibaby committed:
Rename the StretchType enum in Actor

The lowercase enum is confusing because it looks like a variable. This prefixes it with a class name so it's easier to understand its purpose.

sukibaby committed:
Indicate fallthrough to prevent a GCC warning

Without this here, GCC 11 and 12 will print a warning during compilation about a possible fallthrough, but we're doing that on purpose, so this simply adds an indicator to the compiler that we know what we're doing.

sukibaby commented:
Thanks for the review @teejusb !

There is a new commit now (eb9a6d5) which handles the `ActorMultiVertex` warning mentioned in the PR description, since I reverted the `ActorMultiVertex` changes that were originally here.

teejusb commented on src/Actor.h:
```diff
@@ -499,14 +499,14 @@ class Actor : public MessageSubscriber
 
 	/** @brief How do we handle stretching the Actor? */
-	enum StretchType
+	enum class StretchType
```
No need to make it an enum class. Just an enum is fine.

aeubanks replied:
why not enum class? for better name scoping and a little type safety

(I'm not a fan of the `k` prefix but I see the Google C++ style guide says to do that...)

teejusb replied:
Actually I'm trying to think why I suggested not using an enum class and I'm coming up blank. I think maybe familiarity, but I think yeah enum class is fine.

sukibaby replied:
😋

teejusb commented on src/ActorMultiVertex.cpp:
```diff
@@ -502,6 +502,9 @@ void ActorMultiVertex::UpdateAnimationState(bool force_update)
 					break;
 				}
 			}
+#ifdef __GNUC__
```
You shouldn't need this ifdef. It's part of the c++17 spec and we use it in other parts of the codebase.

Also might be worth to see which cases we're missing and just add the labels instead of falling through.

teejusb commented on src/Actor.h:
```diff
@@ -499,14 +499,14 @@ class Actor : public MessageSubscriber
-	enum StretchType
+	enum class StretchType
 	{
 		fit_inside, /**< Have the Actor fit inside its parent, using the smaller zoom. */
```
I'd also propose changing the naming to follow google style guide, which would be:
`kFitInside` and `kCover`

sukibaby committed:
put the enum class back

sukibaby commented:
Feel free to squash and merge btw, as the original post is still accurate.
